Setup databox action

moxcodes opened the pull request:
## Proposed changes

This implements a simple optimization to the construction of the `DataBox` in `Initialization` phases, so that (to the extent possible) tags are added to the `DataBox` in as few places as possible. This limits the number of different databox types that need to be instantiated, and, depending on the compiler, can significantly reduce RAM and compile-time for heavier executables.

Below is the summary of some quick tests that I ran. The take-aways are that maximum RAM usage reduced to ~75% of original for clang-6.0, ~80% for clang-10, and to ~50% of original for GCC-7. Compile times are also improved: ~90% of original compile time for clang, ~35% of original for GCC-7. Clang still outperforms gcc, but by a less dramatic margin. Importantly, I think this will qualitatively improve the system resources requirements for the CI builds.
```
Before:
  Clang-6:
    EvolveValenciaDivCleanAlfvenWave:
      Command being timed: "make -j6 EvolveValenciaDivCleanAlfvenWave"
	   Percent of CPU this job got: 329%
	   Elapsed (wall clock) time (h:mm:ss or m:ss): 3:32.34
	   Maximum resident set size (kbytes): 11164788
    EvolveGhKerrSchild:
      Command being timed: "make -j6 EvolveGhKerrSchild"
	   Percent of CPU this job got: 329%
	   Elapsed (wall clock) time (h:mm:ss or m:ss): 3:05.29
	   Maximum resident set size (kbytes): 8170388
  Clang-10:
    EvolveValenciaDivCleanAlfvenWave:
      Command being timed: "make -j6 EvolveValenciaDivCleanAlfvenWave"
	   Percent of CPU this job got: 348%
	   Elapsed (wall clock) time (h:mm:ss or m:ss): 3:33.99
	   Maximum resident set size (kbytes): 9580212
    EvolveGhKerrSchild:
      Command being timed: "make -j6 EvolveGhKerrSchild"
	   Percent of CPU this job got: 349%
	   Elapsed (wall clock) time (h:mm:ss or m:ss): 3:05.56
	   Maximum resident set size (kbytes): 7223804
  GCC-7:
    EvolveValenciaDivCleanAlfvenWave:
      Command being timed: "make -j6 EvolveValenciaDivCleanAlfvenWave"
	   Percent of CPU this job got: 126%
	   Elapsed (wall clock) time (h:mm:ss or m:ss): 32:49.62
	   Maximum resident set size (kbytes): 27467120
    EvolveGhKerrSchild:
      Command being timed: "make -j6 EvolveGhKerrSchild"
	   Percent of CPU this job got: 135%
	   Elapsed (wall clock) time (h:mm:ss or m:ss): 22:20.10
	   Maximum resident set size (kbytes): 19597960

After:
  Clang-6:
    EvolveValenciaDivCleanAlfvenWave:
      Command being timed: "make -j6 EvolveValenciaDivCleanAlfvenWave"
	   Percent of CPU this job got: 347%
	   Elapsed (wall clock) time (h:mm:ss or m:ss): 3:18.24
	   Maximum resident set size (kbytes): 7761604
    EvolveGhKerrSchild:
      Command being timed: "make -j6 EvolveGhKerrSchild"
	   Percent of CPU this job got: 341%
	   Elapsed (wall clock) time (h:mm:ss or m:ss): 2:59.63
	   Maximum resident set size (kbytes): 6329492
  Clang-10:
    EvolveValenciaDivCleanAlfvenWave:
      Command being timed: "make -j6 EvolveValenciaDivCleanAlfvenWave"
	   Percent of CPU this job got: 362%
	   Elapsed (wall clock) time (h:mm:ss or m:ss): 3:15.83
	   Maximum resident set size (kbytes): 7240564
    EvolveGhKerrSchild:
      Command being timed: "make -j6 EvolveGhKerrSchild"
	   Percent of CPU this job got: 360%
	   Elapsed (wall clock) time (h:mm:ss or m:ss): 2:53.72
	   Maximum resident set size (kbytes): 6008284
  GCC-7:
    EvolveValenciaDivCleanAlfvenWave:
      Command being timed: "make -j6 EvolveValenciaDivCleanAlfvenWave"
	   Percent of CPU this job got: 200%
	   Elapsed (wall clock) time (h:mm:ss or m:ss): 8:46.20
	   Maximum resident set size (kbytes): 12538676
    EvolveGhKerrSchild:
      Command being timed: "make -j6 EvolveGhKerrSchild"
	   Percent of CPU this job got: 198%
	   Elapsed (wall clock) time (h:mm:ss or m:ss): 7:55.48
	   Maximum resident set size (kbytes): 10248700
```

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

moxcodes committed:
Use SetupDataBox with all current initializations

moxcodes committed:
Add db::mutate_assign

moxcodes committed:
Add SetupDataBox action

nilsvu commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1708,7 +1708,7 @@ SPECTRE_ALWAYS_INLINE constexpr auto mutate_apply(
  */
 template <typename F, typename BoxTags, typename... Args>
-SPECTRE_ALWAYS_INLINE constexpr auto mutate_apply(
+SPECTRE_ALWAYS_INLINE constexpr void mutate_apply(
```
The other `mutate_apply` overloads can return, is this overload perhaps just missing the `return`?

nilsvu commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1718,6 +1718,52 @@ SPECTRE_ALWAYS_INLINE constexpr auto mutate_apply(
+  db::mutate<MutateTags...>(
+      box,
+      [&assign, &args...](const gsl::not_null<
+                          typename MutateTags::type*>... box_args) noexcept {
```
Make this `const auto... box_args`, that should be all that's necessary so this function also supports base tags, unless I'm missing something. Please add a test where one of the `MutateTags` is a base tag.

nilsvu commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1718,6 +1718,52 @@ SPECTRE_ALWAYS_INLINE constexpr auto mutate_apply(
+ * \brief Perform a mutation to the \ref DataBoxGroup `box`, assigning the
+ * `args` to the tags `MutateTags` in order.
+ */
+template <typename... MutateTags, typename BoxTags, typename... Args>
```
The `mutate_apply` function takes a `tmpl::list` and the `mutate` function takes a parameter pack of tags (not sure why). I suggest you use the `tmpl::list` approach for this function since I think it's a bit more flexible. What do you think?

Edit: I see you have an overload below that takes a list as a meta argument. I think the interface is better when the function takes the list as template parameter.

moxcodes replied:
I chose to keep an interface with a pack rather than a list so that it mirrored as much as possible the `db::mutate` function. Admittedly, it is a bit awkward that the list interface then needs to be a meta function argument.

If you don't think the analogy with the other utility is enough of a reason to keep this as-is, I can switch over to a pure-list interface.

nilsvu replied:
I see. `db::mutate` is actually inconsistent with `db::mutate_apply` (which does pretty much the same thing) and `db::apply`. @kidder could you weigh in here how you think this interface should look like, since you have been restructuring it recently?

nilsvu replied:
The main reason I think `db::mutate_assign` taking a typelist as template parameter would be good is that all the initialization actions use it with their `simple_tags` list. So the interface that takes a typelist is used a lot. You could make the interface mirror `db::mutate_apply` instead of `db::mutate` (see #2568).

kidder replied:
yes, I'll take a look

kidder replied:
I think making `db::mutate_assign` take a typelist to be consistent with `db::mutate_apply` makes sense.  I think `db::mutate` takes a pack to make it more convenient for mutating one tag

nilsleiffischer replied:
ping on this thread. If I understand @kidder correctly, the interface both of us are suggesting is:

```cpp
template <typename MutateTags, typename BoxTags, typename... Args>
SPECTRE_ALWAYS_INLINE constexpr void mutate_assign(
     const gsl::not_null<DataBox<BoxTags>*> box, Args&&... args) noexcept;
```

This is consistent with `mutate_apply` in https://github.com/sxs-collaboration/spectre/pull/2568/.

nilsvu replied:
ping on this thread. If I understand @kidder correctly, the interface both of us are suggesting is:

```cpp
template <typename MutateTags, typename BoxTags, typename... Args>
SPECTRE_ALWAYS_INLINE constexpr void mutate_assign(
     const gsl::not_null<DataBox<BoxTags>*> box, Args&&... args) noexcept;
```

This is consistent with `mutate_apply` in https://github.com/sxs-collaboration/spectre/pull/2568/.

nilsvu reviewed (changes requested):
Only reviewed the first commit so far. Thanks for doing this @moxcodes!

nilsvu commented on src/Evolution/Initialization/ConservativeSystem.hpp:
```diff
@@ -63,29 +65,53 @@ namespace Actions {
-
     const size_t num_grid_points =
         db::get<domain::Tags::Mesh<dim>>(box).number_of_grid_points();
+    Parallel::printf("grrr %zu\n", num_grid_points);
```
In the spirit of not frightening the user I suggest to remove this message ;)

moxcodes replied:
oops -- a remnant from a particularly frustrating debugging session :P

nilsvu commented on src/Parallel/Actions/SetupDataBox.hpp:
```diff
@@ -0,0 +1,132 @@
+#include "Parallel/GlobalCache.hpp"
+#include "ParallelAlgorithms/Initialization/MergeIntoDataBox.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
```
Include `Utilities/TypeTraits/CreateHasTypeAlias.hpp`

nilsvu commented on src/Parallel/Actions/SetupDataBox.hpp:
```diff
@@ -0,0 +1,132 @@
+}  // namespace detail
+
+/*!
+ * \brief Merge into the \ref DataBoxGroup the collection of tags requested for
```
You can write `\ref DataBoxGroup "DataBox"` to make sure the link renders as "DataBox" (instead of the title of the linked page) and for easier reading of the non-rendered docs

nilsvu commented on src/Elliptic/Actions/InitializeAnalyticSolution.hpp:
```diff
@@ -41,10 +41,20 @@ namespace Actions {
+ *
+ * \note This action relies on the `SetupDataBox` aggregated initialization
+ * mechanism, so `Actions::SetupDataBox` must be present in the `Initialization`
+ * phase action list prior to this action.
```
This currently fairly "loose" documentation-based link between the initialization actions and the `SetupDataBox` mechanism seems like a prime example for a use case of a protocol :) Here's my suggestion to more strongly define and document the compile-time interface for these "initialization actions":

1. Create the following protocol in `Parallel/Protocols.hpp`:

    ```cpp
    /*!
     * \brief An action that initializes tags in the DataBox
    *
    * Actions that conform to this protocol rely on the `SetupDataBox` aggregated initialization
    * mechanism. They must provide the typelists `simple_tags` and `compute_tags`. Tags in these
    * typelists are added to the DataBox by the `SetupDataBox` action, which must be present
    * in the phase action-list prior to the initialization action. The initialization action can then
    * mutate the tags to set their initial values.
    *
    * Requires the conforming action has these type aliases:
    * - `simple_tags`: A `tmpl::list` of `db::SimpleTag`s that will be added to the DataBox and
    * initially default-constructed. The initialization action can mutate these tags to set their initial values.
    * - `compute_tags`: A `tmpl::list` of `db::ComputeTag`s that will be added to the DataBox.
    *
    * Here's an example of an action that conforms to this protocol:
    * \snippet ...
    */
    struct InitializationAction {
      template <typename ConformingType>
      struct test {
        using simple_tags = typename ConformingType::simple_tags;
        using compute_tags = typename ConformingType::compute_tags;
        // We could also assert that `simple_tags` contains no compute tags, and vice-versa
      };
    };
    ```
    Add the example snippet to `tests/Unit/Parallel/Test_Protocols.cpp` (see `tests/Unit/Evolution/Test_Protocols.cpp` for an example).
2. In `SetupDataBox`, instead of checking whether each action has `simple_tags` and `compute_tags` aliases, check `tt::conforms_to_v<Parallel::protocols::InitializationAction>`. When this is true, you can rely on the action having the aliases.
3. Derive all initialization actions from `tt::ConformsTo<Parallel::protocols::InitializationAction>`. Also add to their tests the line:
    ```cpp
    static_assert(tt::assert_conforms_to<TheAction, Parallel::protocols::InitializationAction>)`
    ```

Now the protocol defines what an "initialization action" is, provides documentation and gives an example.
Only actions that indicate they are initialization actions by conforming to the protocol are picked up by `SetupDataBox`. This establishes the strong link from the action to the documentation of what it means to be an initialization action, and makes it unit-testable that the action will work correctly with `SetupDataBox`. Let me know what you think :)

moxcodes replied:
I think that type of constraint can be useful, but some (particularly @nilsdeppe) have expressed reservations about extensive usage of the protocols because of challenges that can arise during debugging. Because this constraint would affect every initialization action (which can already be difficult to debug for other reasons), I'd like another voice in favor of using the protocols for this before implementing the change everywhere.
If we don't apply this immediately for this PR, we can open an issue indicating that this would be a good place to consider for protocols.

nilsdeppe replied:
I haven't tested the newer protocol revamps extensively, partly because I've been avoiding using protocols after my last experience. I would be mildly against making this switch because (a) I'm not yet completely convinced protocols work well enough and (b) it's a lot of work for something we don't know if it works well enough. I don't disagree that having some checks like what protocols offers (and the associated documentation) would be useful, but I'm hesitant to suggest that a +1600/-1000 line PR is the place to test this... I will say that you should _not_ put in in a `Protocols.hpp`. Having a single file to collect all protocols will not scale. We've tried that with tags and it didn't work.

nilsvu replied:
That's fair. I think this would be a very fitting use case for a protocol, but if you don't want to go that route for the reasons @nilsdeppe gave that's totally fine with me.

moxcodes replied:
I've made issue #2570 for now to list protocols to add once protocols are more thoroughly tested elsewhere.

nilsvu replied:
Sounds good. I added a very limited use of protocols to #2569 where it makes a lot of sense as well but doesn't affect such a large part of our codebase :)

nilsvu commented on src/Parallel/Actions/SetupDataBox.hpp:
```diff
@@ -0,0 +1,132 @@
+    // grab the simple_tags, compute_tags, mutate the databox, creating
+    // default-constructed objects.
+    return std::make_tuple(detail::merge_into_databox_helper(
+        std::move(box), action_list_simple_tags{}, action_list_compute_tags{}));
```
I occasionally encounter #2524 when merging lots of simple tags into a DataBox at once. I haven't investigated the issue in much detail, so I'm not sure if it's the number of tags or the amount of data that's relevant. If it's the number of tags, then `SetupDataBox` has a high likelihood of triggering the issue. If it's the amount of data, then the new initialization mechanism might actually fix it. Have you seen the issue at all? Any ideas what could cause it?

moxcodes replied:
I haven't encountered that error -- are there any steps you can suggest to reliably reproduce it? When you encountered it, was it locally in the container, or on a particular HPC system?

nilsdeppe replied:
I've also never encountered that issue... The issue is quite sparse. There are no details about compiler, env, etc. Please fill that out otherwise we can't help and should just close the issue.

nilsvu replied:
I encountered it on my laptop, trying to merge ~30 simple tags containing tensors (maybe more) into the DataBox at once. I haven't tried reproducing it in a controlled environment yet, but that would probably worthwhile.

nilsvu replied:
Here's a patch to current `develop` that reproduces the issue on my machine: https://github.com/nilsleiffischer/spectre/commit/d59ac5b6ac1735ded441d13a3c0373c26146ca76

```
$ make -j12 ExportCoordinates3D \
    && ./bin/ExportCoordinates3D +p2 \
      --input-file tests/InputFiles/ExportCoordinates/Input3D.yaml

[1]    22117 bus error  ./bin/ExportCoordinates3D --input-file  +p2
```

nilsdeppe replied:
Umm 300 is _a lot_. Do you know roughly where this breaks for you?

moxcodes replied:
sorry @nilsleiffischer, I wasn't able to reproduce the problem either locally or on wheeler with your patch -- the sigbus makes me think it is related to some hardware based oddity, so this might need to be investigated by someone who has a macintosh laptop to test on.

nilsvu replied:
Thanks for checking @moxcodes. Since I also can't reproduce the problem in the Docker container we should just move on, and I'll need to track this down further. Below are some more details, but let's not hold up this PR with this discussion.

---

In this patch to `ExportCoordinates` it breaks somewhere between 200 and 300 tags when running on 2 cores, compiled on macOS with AppleClang 11 or 12. In the XCTS executable it breaks when adding a `Variables` with only 15 tags to the DataBox (the background variables such as conformal metric, energy density etc), and splitting that call into 3 separate `db::create_from`s resolves it. This makes me think it's somehow related to system load, or the size of the DataBox that is being added to. I tracked it down to this line
https://github.com/sxs-collaboration/spectre/blob/d1ac7e2676842c20ecc43fa2107122187d06d254/src/DataStructures/DataBox/DataBox.hpp#L1287
where the `new_tag_list` is a list of all tags, including the subitems from the Variables, which can be a lot. The `DataBox` constructor doesn't appear to get called, so I wonder if the issue might be in the way the DataBox allocates memory.. perhaps related to the inheritance off so many `DataBoxLeaf`s..

nilsdeppe replied:
this sounds like a compiler bug in AppleClang. It's not that we are inheriting too many tags. Even the Intel compiler can handle inheriting many tags. Additionally, if that were the issue it would be a problem independent of how you construct the final `DataBox`. Can you test LLVM clang on macOS? It might be that we need to drop AppleClang

nilsvu replied:
Compiling with either llvm clang 10 or 11 I get a segfault instead of the sigbus.. 🤔

nilsdeppe replied:
on the same line? segfaults are usually easier to diagnose than sigbus, but it's not clear that will hold here :P

nilsvu replied:
yes on the same line :(

moxcodes commented:
fixups posted.
Cheers!

moxcodes commented:
... and a couple of force-pushes to fix weird clang-tidy complaints
Cheers!

nilsvu commented on src/Parallel/Actions/SetupDataBox.hpp:
```diff
@@ -0,0 +1,132 @@
+CREATE_HAS_TYPE_ALIAS(compute_tags)
+CREATE_HAS_TYPE_ALIAS_V(compute_tags)
+
+template <typename Action, typename enable = std::bool_constant<true>>
```
You can do this without `has_type_alias`. In this line, replace the `enable` parameter with `typename = std::void_t<>` and immediately define this struct with the `type = tmpl::list<>`. Then, in the specialization below, replace the `std::bool_constant` with `std::void_t<typename Action::simple_tags>`. SFINAE will do the right thing :) The same applies to the compute tags of course.

nilsvu commented on src/Parallel/Actions/SetupDataBox.hpp:
```diff
@@ -0,0 +1,132 @@
+template <typename... ActionList>
+struct get_action_list_simple_tags<tmpl::list<ActionList...>> {
+  using type = tmpl::remove_duplicates<tmpl::flatten<
+      tmpl::list<typename optional_simple_tags<ActionList>::type...>>>;
```
I think this is just `tmpl::transform<ActionList, tmpl::bind<optional_simple_tags, tmpl::_1>>`. Then you can make `get_action_list_simple_tags` a type alias instead of a struct.

nilsvu commented on src/Parallel/Actions/SetupDataBox.hpp:
```diff
@@ -0,0 +1,132 @@
+  if constexpr (not tmpl2::flat_any_v<
+                    tmpl::list_contains_v<DbTags, simple_tags>...> and
+                not tmpl2::flat_any_v<
+                    tmpl::list_contains_v<DbTags, compute_tags>...>) {
```
I think the check that none of the compute tags already exist in the DataBox is not necessary, since `db::create_from` merges the compute tags anyway. The check makes the `SetupDataBox` action a bit more restrictive than it needs to be, so I suggest you remove it.

nilsvu replied:
ping, in case you missed this

moxcodes replied:
ah, sorry, forgot to comment -- this didn't work. My understanding of the error I encountered was an edge case associated with the base tags of the compute tags and logic involving multiple passes through the initialization phase (which we don't usually use, but must always compile correctly because the initialization phase still 'loops').

nilsvu replied:
Do you remember what specifically didn't work? Adding a compute tag multiple times should work (and just do nothing). The main reason I'm suggesting to remove this restriction is that it can break in some edge-cases, e.g. when the `SetupDataBox` is not the very first action and the DataBox already has some compute tags in it for whatever reason.

moxcodes replied:
I think you must be recalling functionality that has since been removed in more recent `DataBox` changes. In particular, on develop the following patch that uses `db::create_from` with an existing compute tag causes compilation failure:
```
diff --git a/tests/Unit/DataStructures/DataBox/Test_DataBox.cpp b/tests/Unit/DataStructures/DataBox/Test_DataBox.cpp
index da4a61866..955a5bf49 100644
--- a/tests/Unit/DataStructures/DataBox/Test_DataBox.cpp
+++ b/tests/Unit/DataStructures/DataBox/Test_DataBox.cpp
@@ -237,7 +237,11 @@ void test_databox() noexcept {
                            test_databox_tags::Lambda1Compute>>(
         3.14, std::vector<double>{8.7, 93.2, 84.7}, "My Sample String"s);
     /// [create_databox]
-    return original_box;
+    auto new_box =
+        db::create_from<db::RemoveTags<>, db::AddSimpleTags<>,
+                        db::AddComputeTags<test_databox_tags::Tag4Compute>>(
+            std::move(original_box));
+    return new_box;
   };
   {
     const auto original_box = create_original_box();
```
Or perhaps are you thinking of the `initialization::merge_into_databox` utility? I didn't use that because it is more restrictive in requiring equivalence operators for the tagged types.

nilsvu replied:
You're right, only `merge_into_databox` resolves duplicate compute items. I suggest you do the same here, adding only compute items that are not already present in the DataBox. This is the relevant snippet from `MergeIntoDataBox.hpp`:

```cpp
  using compute_tags_to_add = tmpl::remove_if<
      ComputeTagsList,
      tmpl::bind<tmpl::list_contains,
                 tmpl::pin<typename db::DataBox<DbTagsList>::compute_item_tags>,
                 tmpl::_1>>;
```

wthrowe replied:
I'm not pushing for the functionality change, but I want to point out that this is more simply written as `tmpl::list_difference<ComputeTagsList, typename db::DataBox<DbTagsList>::compute_item_tags>`.

moxcodes replied:
I've added another fixup that performs the change for both compute tags and for simple tags -- now both have 'merge' behavior, where they'll just add tags that aren't already present.

nilsvu commented on src/Parallel/Actions/SetupDataBox.hpp:
```diff
@@ -0,0 +1,132 @@
+      tmpl::list<typename optional_compute_tags<ActionList>::type...>>>;
+};
+
+template <typename DbTags, typename... simple_tags, typename... compute_tags>
```
Please make template parameters CamelCase

nilsvu commented on tests/Unit/Parallel/Actions/Test_SetupDataBox.cpp:
```diff
@@ -0,0 +1,96 @@
+};
+
+
+struct InitializationAction {
```
How about you add this example to the documentation of `SetupDataBox` as a `\snippet`?

nilsvu commented on tests/Unit/Parallel/Actions/Test_SetupDataBox.cpp:
```diff
@@ -0,0 +1,96 @@
+  }
+};
+
+/// [component]
```
I don't see you use these snippets in the docs, so just remove the comments

nilsvu reviewed (changes requested):
I'm happy with the fixups. You could squash additional changes into them before a second reviewer looked at them.

moxcodes commented:
squashed in the additional changes to the existing fixups.
Cheers!

nilsvu commented:
LGTM! Please squash.

moxcodes commented:
rebased on develop to fix conflicts.

moxcodes commented:
fixups squashed in.
Cheers!

nilsvu reviewed (commented):
I'm very much looking forward to trying this with my XCTS executable :D looks super promising

moxcodes commented:
fixups posted for the `mutate_assign` interface change.
Cheers!

nilsvu commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1718,54 +1718,39 @@ SPECTRE_ALWAYS_INLINE constexpr auto mutate_apply(
         // silence unused capture warnings
         (void)assign;
         EXPAND_PACK_LEFT_TO_RIGHT((void)args);
         EXPAND_PACK_LEFT_TO_RIGHT(assign(box_args, std::forward<Args>(args)));
```
I think you can do:

```cpp
expand_pack((*box_args = std::forward<Args>(args))...);
```

moxcodes replied:
Ah, yeah, I noticed that on your PR. I'll switch to that, though I think I'll keep the order preservation because I believe databox assignment can actually depend on order sometimes (e.g. subtags).

nilsvu commented:
The interface looks great now IMO, I'll look through the initializers again tomorrow. Thanks for your work on this!

moxcodes commented:
the last mutate_assign simplification squashed in to `fixup: db::mutate_assign`
Cheers!

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1715,7 +1715,36 @@ SPECTRE_ALWAYS_INLINE constexpr auto mutate_apply(
       "'return_tags'. Did you forget to add these tag lists to the mutator "
       "class? The class is listed as the first template parameter below.");
-  mutate_apply(F{}, box, std::forward<Args>(args)...);
+  return mutate_apply(F{}, box, std::forward<Args>(args)...);
```
I've changed this function's return type to void in #2568 as it cannot return anything (it eventually calls `db::mutate` which returns void)

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1715,7 +1715,36 @@ SPECTRE_ALWAYS_INLINE constexpr auto mutate_apply(
+                "The number of arguments passed to `mutate_assign` must be "
+                "equal to the number of tags passed.");
+  db::mutate<MutateTags...>(box, [&args...](const auto... box_args) noexcept {
+    // silence unused capture warnings
```
when do you get unused warning?  If it is only for an empty list of `MutateTags` then do we want to allow an empty list or should there be a static_assert that the list has at least one element?

nilsdeppe replied:
ping

moxcodes replied:
For compiler reasons I do not entirely understand, this warns even when I have ensured that nothing that calls it does so with 0 tags.

nilsdeppe replied:
oh, it's because it's still being instantiated... :( How about just adding to the comment `when there are zero args. This function still gets instantiated despite the static_assert in the parent function.`

kidder commented on tests/Unit/DataStructures/DataBox/Test_DataBox.cpp:
```diff
@@ -1400,9 +1400,12 @@ void test_mutate_apply() noexcept {
         make_not_null(&box));
 
     struct PointerMutateApply {
+#pragma GCC diagnostic push
```
I fixed these in #2568 by moving the structs outside of the function

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1715,7 +1715,36 @@ SPECTRE_ALWAYS_INLINE constexpr auto mutate_apply(
+
+/*!
+ * \ingroup DataBoxGroup
+ * \brief Perform a mutation to the \ref DataBoxGroup `box`, assigning the
```
since this function should only be uses during initialization, let's move it from `DataBox.hpp` to somewhere in the `Initialization` namespace

nilsvu replied:
This is just a DataBox utility, not inherently related to initialization. I don't see a reason we should discourage using this function. For example, I can see it be useful to move data from an inbox into the DataBox in a receive action.

nilsdeppe replied:
Yea, we don't want anyone to ever do that... Basically, this utility will cause people to blindly never reuse memory allocations and then we have to waste time fixing lots of code. It's better to prevent a bug than to fix one.

nilsvu replied:
I'm not sure I understand - how else would you move data from the inbox into the DataBox?

nilsdeppe replied:
`db::mutate`. My concern is that people will start using `mutate_assign` everywhere, and in many places where you should be using `db::mutate`. The "it's less code I'll just use that" issue. This will mean we don't use re-use a lot of memory allocations and just adds more manual policing to the code review process. The number of times `mutate_assign` should be used outside of initialization is close to zero, so I'd rather eliminate an entire class of bugs than have to add yet another thing to the "Please remember to check this while reviewing code" list.

nilsvu replied:
ok

kidder commented on src/Parallel/Actions/SetupDataBox.hpp:
```diff
@@ -0,0 +1,106 @@
+    tmpl::transform<ActionList, optional_compute_tags<tmpl::_1>>>>;
+
+template <typename DbTags, typename... SimpleTags, typename... ComputeTags>
+static auto merge_into_databox_helper(
```
remove `static`

kidder commented on src/Parallel/Actions/SetupDataBox.hpp:
```diff
@@ -0,0 +1,106 @@
+}  // namespace detail
+
+/*!
+ * \brief Merge into the \ref DataBoxGroup "DataBox" the collection of tags
```
Add into the DataBox default constructed items for the collection of tags requested by any of the actions in the current phase.

kidder commented on tests/Unit/Parallel/Actions/Test_SetupDataBox.cpp:
```diff
@@ -0,0 +1,94 @@
+      const Parallel::GlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& /*array_index*/, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    CHECK(db::get<CounterTag>(box) == 0_st);
```
since this is a snippet, it might be better to move the `CHECK`s out of the action and do them before and after the action is called

kidder replied:
please add a second action that mutates the counter.

kidder commented on tests/Unit/Parallel/Actions/Test_SetupDataBox.cpp:
```diff
@@ -0,0 +1,94 @@
+  using MockRuntimeSystem = ActionTesting::MockRuntimeSystem<Metavariables>;
+  MockRuntimeSystem runner{{}};
+  ActionTesting::emplace_component<component>(&runner, 0);
+  for (size_t i = 0; i < 2; ++i) {
```
then call both actions and check all items in the DataBox between calls

kidder commented on src/Elliptic/Actions/InitializeSystem.hpp:
```diff
@@ -43,22 +43,31 @@ namespace Actions {
+ * mechanism, so `Actions::SetupDataBox` must be present in the `Initialization`
+ * phase action list prior to this action.
  */
+template <typename Metavariables>
```
We would like to avoid having actions templated on the Metavariables.  In this case, how about templating on `System`

kidder commented on src/Evolution/Initialization/ConservativeSystem.hpp:
```diff
@@ -25,6 +25,8 @@
 #include "Utilities/Requires.hpp"
 #include "Utilities/TMPL.hpp"
 
+#include "Parallel/Printf.hpp"
```
RIP

kidder commented on src/Evolution/Initialization/ConservativeSystem.hpp:
```diff
@@ -63,27 +65,50 @@ namespace Actions {
+                                         tmpl::size_t<dim>, Frame::Inertial>;
+  using sources_tag = db::add_tag_prefix<::Tags::Source, variables_tag>;
+
+  template <typename System, typename enable = std::true_type>
```
`bool = System::has_primitive_and_conservative_vars`

kidder commented on src/Evolution/Initialization/ConservativeSystem.hpp:
```diff
@@ -63,27 +65,50 @@ namespace Actions {
+
+  template <typename System>
+  struct simple_tags_impl<
+      System, std::bool_constant<System::has_primitive_and_conservative_vars>> {
```
`std::bool_constant<>` -> true

kidder commented on src/Evolution/Initialization/ConservativeSystem.hpp:
```diff
@@ -63,27 +65,50 @@ namespace Actions {
+/// \note This action relies on the `SetupDataBox` aggregated initialization
+/// mechanism, so `Actions::SetupDataBox` must be present in the
+/// `Initialization` phase action list prior to this action.
+template <typename Metavariables>
```
template on `typename System, typename EquationOfStateTag = NoSuchType` instead of Metavariables

kidder commented on src/Evolution/Initialization/GrTagsForHydro.hpp:
```diff
@@ -57,23 +57,28 @@ namespace Actions {
+/// \note This action relies on the `SetupDataBox` aggregated initialization
+/// mechanism, so `Actions::SetupDataBox` must be present in the
+/// `Initialization` phase action list prior to this action.
+template <typename Metavariables>
```
template on System

kidder commented on src/Evolution/Initialization/NonconservativeSystem.hpp:
```diff
@@ -39,28 +39,33 @@ namespace Actions {
+/// \note This action relies on the `SetupDataBox` aggregated initialization
+/// mechanism, so `Actions::SetupDataBox` must be present in the
+/// `Initialization` phase action list prior to this action.
+template <typename Metavariables>
```
template on System

kidder commented on src/Evolution/Systems/Cce/Actions/InitializeCharacteristicEvolutionScri.hpp:
```diff
@@ -35,65 +35,54 @@ namespace Actions {
+ * mechanism, so `Actions::SetupDataBox` must be present in the `Initialization`
+ * phase action list prior to this action.
  */
+template <typename Metavariables>
```
could we template this on ScriValuesToObserve?

kidder commented on src/Evolution/Systems/Cce/Actions/InitializeCharacteristicEvolutionTime.hpp:
```diff
@@ -52,34 +52,36 @@ namespace Actions {
+ * mechanism, so `Actions::SetupDataBox` must be present in the `Initialization`
+ * phase action list prior to this action.
  */
+template <typename Metavariables>
```
again template on the two things grabbed from the Metavariables

kidder commented on src/Evolution/Systems/Cce/Actions/InitializeCharacteristicEvolutionVariables.hpp:
```diff
@@ -100,6 +93,27 @@ struct InitializeCharacteristicEvolutionVariables {
     using evolved_swsh_dt_variables_tag =
         db::add_tag_prefix<::Tags::dt, evolved_swsh_variables_tag>;
 
+  using simple_tags = tmpl::list<
```
not for this PR, but something to consider is to pack all of these things defined in `Metavariables` be defined in a struct `CceOptions` that is then in the `Metavariables` as `Metavariables::cce_options`

moxcodes replied:
I agree.. now that the CCE system is mostly complete and I have a clearer view of it, I think there's a bunch of tag lists that I'd like to aggregate into a clear central location like that.

kidder commented on src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp:
```diff
@@ -44,7 +44,12 @@ namespace Actions {
+ * mechanism, so `Actions::SetupDataBox` must be present in the `Initialization`
+ * phase action list prior to this action.
  */
+template <typename Metavariables>
```
template on the communication tags

kidder commented on src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp:
```diff
@@ -118,7 +114,12 @@ struct InitializeH5WorldtubeBoundary {
+ * mechanism, so `Actions::SetupDataBox` must be present in the `Initialization`
+ * phase action list prior to this action.
  */
+template <typename Metavariables>
```
template on communication tags

kidder commented on src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp:
```diff
@@ -41,25 +41,25 @@ namespace RadiationTransport {
 namespace M1Grey {
 namespace Actions {
 
+template <typename Metavariables>
```
template on System

kidder commented on src/IO/Importers/ElementDataReader.hpp:
```diff
@@ -57,6 +59,10 @@ struct ElementDataReader {
 namespace detail {
 struct InitializeElementDataReader {
+  using simple_tags =
+      db::AddSimpleTags<Tags::RegisteredElements, Tags::ElementDataAlreadyRead>;
```
you've been using `tmpl::list` everywhere instead of `db::AddSimpleTags`; please be consistent

kidder commented on src/NumericalAlgorithms/Interpolation/Actions/ElementInitInterpPoints.hpp:
```diff
@@ -37,36 +37,22 @@ namespace Actions {
+/// \note This action relies on the `SetupDataBox` aggregated initialization
+/// mechanism, so `Actions::SetupDataBox` must be present in the
+/// `Initialization` phase action list prior to this action.
+template <typename Metavariables>
```
template on `InterpPointInfoTag`

kidder commented on src/NumericalAlgorithms/Interpolation/Actions/ElementInitInterpPoints.hpp:
```diff
@@ -37,36 +37,22 @@ namespace Actions {
-              db::AddSimpleTags<intrp::Tags::InterpPointInfo<Metavariables>>>(
-              std::move(box), point_info_type{}));
-    }
+    return std::make_tuple(std::move(box));
```
add a comment that the default initialization done by `SetupDataBox` was sufficient.

kidder commented on src/NumericalAlgorithms/Interpolation/InitializeInterpolationTarget.hpp:
```diff
@@ -60,21 +60,37 @@ template <
+    const Parallel::GlobalCache<Metavariables>& /*cache*/) noexcept {}
+
+template <typename InterpolationTargetTag, typename = std::void_t<>,
+          typename = std::bool_constant<true>>
```
please just use `bool  = CHECK_TO_PERFORM` instead of `typename = std::bool_consant<>`

kidder commented on src/NumericalAlgorithms/Interpolation/InitializeInterpolationTarget.hpp:
```diff
@@ -60,21 +60,37 @@ template <
         nullptr>
-auto make_initial_box(
-    db::DataBox<DbTags>&& box,
+void make_initial_box(
```
make_initial_box -> mutate_initial_box

kidder commented on src/NumericalAlgorithms/Interpolation/InitializeInterpolator.hpp:
```diff
@@ -45,51 +45,30 @@ namespace Actions {
+/// \note This action relies on the `SetupDataBox` aggregated initialization
+/// mechanism, so `Actions::SetupDataBox` must be present in the
+/// `Initialization` phase action list prior to this action.
+template <typename Metavariables>
```
template on the two tags instead of the Metavariables

kidder commented on src/ParallelAlgorithms/DiscontinuousGalerkin/InitializeInterfaces.hpp:
```diff
@@ -49,6 +49,17 @@ using exterior_compute_tags = tmpl::list<Tags...>;
 namespace Actions {
 
+namespace detail {
+template <typename System, typename enable = std::void_t<>>
```
you don't need to specify `enable` (i.e it's an unnamed template parameter)

kidder commented on src/ParallelAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitor.hpp:
```diff
@@ -49,7 +51,9 @@ struct ResidualMonitor {
-          tmpl::list<InitializeResidualMonitor<FieldsTag, OptionsGroup>>>,
+          tmpl::list<::Actions::SetupDataBox,
+                     InitializeResidualMonitor<FieldsTag, OptionsGroup>,
+                     Initialization::Actions::RemoveOptionsAndTerminatePhase>>,
```
why is this added?

kidder commented on tests/Unit/Framework/ActionTesting.hpp:
```diff
@@ -471,8 +471,10 @@ using item_type_if_contained_t =
-  using compute_tags = ComputeTagsList;
+  // these type aliases must be different from those used by `SetupDataBox` to
+  // avoid a clash.
+  using simple_tags_to_add = tmpl::list<SimpleTags...>;
```
simple_tags_to_add -> action_testing_simple_tags
compute_tags_to_add -> action_testing_compute_tags

in order to avoid potential future name clashes...

kidder commented on tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/LinearSolverAlgorithmTestHelpers.hpp:
```diff
@@ -270,7 +271,7 @@ struct ElementArray {
       Parallel::PhaseActions<
           typename Metavariables::Phase, Metavariables::Phase::Initialization,
-          tmpl::list<InitializeElement,
+          tmpl::list<InitializeElement, Actions::SetupDataBox,
```
why is SetupDataBox after InitializeElement?

kidder commented on tests/Unit/Parallel/Actions/CMakeLists.txt:
```diff
@@ -6,6 +6,7 @@ set(LIBRARY "Test_ParallelActions")
 set(LIBRARY_SOURCES
   Test_Goto.cpp
   Test_TerminatePhase.cpp
+  Test_SetupDataBox.cpp
```
alphabet

kidder commented on tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_FluxCommunicationLts.cpp:
```diff
@@ -159,10 +160,10 @@ struct lts_component {
   using init_mortars_tags =
-      tmpl::list<::Tags::Mortars<::Tags::Next<TemporalIdTag>, Dim>,
-                 ::Tags::Mortars<MortarDataTag<Dim>, Dim>,
+      tmpl::list<::Tags::Mortars<MortarDataTag<Dim>, Dim>,
```
why did you reorder the list?

moxcodes replied:
It needs to be in the same order as the corresponding `simple_tags` from `InitializeMortars`, otherwise the `get_databox` calls in this fail.

kidder commented on tests/Unit/ParallelAlgorithms/LinearSolver/AsynchronousSolvers/Test_ElementActions.cpp:
```diff
@@ -17,7 +17,7 @@
 #include "IO/H5/File.hpp"
 #include "IO/Observer/Helpers.hpp"
 #include "NumericalAlgorithms/Convergence/HasConverged.hpp"
-#include "NumericalAlgorithms/Convergence/Tags.hpp"
```
why was this removed?

kidder replied:
ping

kidder commented on tests/Unit/ParallelAlgorithms/LinearSolver/ConjugateGradient/Test_ElementActions.cpp:
```diff
@@ -13,7 +13,7 @@
 #include "DataStructures/DenseVector.hpp"
 #include "Framework/ActionTesting.hpp"
 #include "NumericalAlgorithms/Convergence/HasConverged.hpp"
-#include "NumericalAlgorithms/Convergence/Tags.hpp"
```
why is this removed?

kidder reviewed (changes requested):
reviewed with @nilsdeppe

moxcodes commented:
Another round of fixups posted (prefixed with `fixup 2:`)
Cheers!

kidder commented on tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/LinearSolverAlgorithmTestHelpers.hpp:
```diff
@@ -270,7 +270,7 @@ struct ElementArray {
       Parallel::PhaseActions<
           typename Metavariables::Phase, Metavariables::Phase::Initialization,
-          tmpl::list<InitializeElement,
+        tmpl::list< Actions::SetupDataBox,InitializeElement,
```
this formatting looks off

kidder reviewed (commented):
@nilsdeppe and I are happy for you to squash the remaining changes
please wait for @nilsleiffischer to also give the go ahead to squash

nilsvu commented on src/ParallelAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitor.hpp:
```diff
@@ -14,10 +14,12 @@
 #include "Parallel/ParallelComponentHelpers.hpp"
 #include "Parallel/PhaseDependentActionList.hpp"
-#include "ParallelAlgorithms/Initialization/MergeIntoDataBox.hpp"
+#include "ParallelAlgorithms/Initialization/Actions/RemoveOptionsAndTerminatePhase.hpp"
```
Please delete this line

nilsvu commented on src/ParallelAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitor.hpp:
```diff
@@ -87,17 +93,12 @@ struct InitializeResidualMonitor {
-            std::numeric_limits<double>::signaling_NaN()),
-        true);
+    // The `InitializeResidual` action populates these
+    // tags with initial values
```
Might as well rewrap this comment when deleting the include above

nilsvu commented on src/ParallelAlgorithms/LinearSolver/ConjugateGradient/InitializeElement.hpp:
```diff
@@ -51,24 +58,18 @@ struct InitializeElement {
+        typename operator_applied_to_fields_tag::type{},
+        typename operand_tag::type{},
+        typename operator_applied_to_operand_tag::type{},
+        typename residual_tag::type{}, Convergence::HasConverged{});
```
Please delete these default-constructions, as you did in `AsynchronousSolvers/ElementActions.hpp`

nilsvu commented on src/ParallelAlgorithms/LinearSolver/Gmres/InitializeElement.hpp:
```diff
@@ -46,8 +46,22 @@ struct InitializeElement {
+      LinearSolver::Tags::KrylovSubspaceBasis<preconditioned_operand_tag>;
 
  public:
+  using simple_tags = tmpl::append<
+      tmpl::list<Convergence::Tags::IterationId<OptionsGroup>,
+                 initial_fields_tag, operator_applied_to_fields_tag,
+                 operand_tag, operator_applied_to_operand_tag,
+                 orthogonalization_iteration_id_tag, basis_history_tag,
+                 Convergence::Tags::HasConverged<OptionsGroup>>,
+      tmpl::conditional_t<Preconditioned,
+                          tmpl::list<preconditioned_basis_history_tag,
+                                     preconditioned_operand_tag>,
+                          tmpl::list<>>>;
```
I remember `tmpl::append` had some bug related to empty lists, @kidder @wthrowe @nilsdeppe can you remind me what the failure mode actually is? I have been using this pattern recently:

```suggestion
  using simple_tags = tmpl::flatten<tmpl::list<
      Convergence::Tags::IterationId<OptionsGroup>,
                 initial_fields_tag, operator_applied_to_fields_tag,
                 operand_tag, operator_applied_to_operand_tag,
                 orthogonalization_iteration_id_tag, basis_history_tag,
                 Convergence::Tags::HasConverged<OptionsGroup>,
      tmpl::conditional_t<Preconditioned,
                          tmpl::list<preconditioned_basis_history_tag,
                                     preconditioned_operand_tag>,
                          tmpl::list<>>>>;
```

wthrowe replied:
The container has a new enough version of Brigand that that bug is fixed, so if it passes CI it's fine.  (See also all the failures on #2326.)

nilsvu replied:
ok then I'm happy with the code as-is

nilsvu reviewed (commented):
Yes please squash, including these minor changes

nilsvu commented:
Probably best to rebase on develop to make sure no issues come up with #2568

moxcodes commented:
rebased and fixups squashed in (along with the last comments)
Cheers!

nilsvu commented on tests/Unit/Framework/ActionTesting.hpp:
```diff
@@ -471,8 +471,10 @@ using item_type_if_contained_t =
-  using simple_tags = tmpl::list<SimpleTags...>;
-  using compute_tags = ComputeTagsList;
+  // these type aliases must be different from those used by `SetupDataBox` to
+  // avoid a clash.
```
I would just like to note here that a protocol would resolve this clash :) The `SetupDataBox` action would only pick up those actions that indicate they conform to the protocol, making the coupling explicit.

Not requesting a change here, I agree with introducing protocols on smaller scales first, just wanted to point this out.

kidder commented:
@moxcodes this needs another rebase

moxcodes commented:
Rebased again. There were more initialization actions that needed to be converted to the new interface, so I added those as a fixup so they're easy to find.
Cheers!

nilsdeppe reviewed (commented):
LGTM, feel free to squash

nilsvu reviewed (dismissed):
yes please squash

kidder reviewed (commented):
you can squash

moxcodes commented:
Squashed.
Cheers!
